### ISSUE TITLE
Expose logits (token-level scores) for text in MediaPipe LLM Inference GenAI Task

### ISSUE THREAD

**Issue by fosple** [google-ai-edge/mediapipe#6124]
### OS Platform and Distribution

Web

### Task name (e.g. Image classification, Gesture recognition etc.)

LLM Inference (tasks-genai)

### Are you willing to contribute it

Yes, I would love to. However, it looks like [WasmLlmInferenceModule](https://github.com/google-ai-edge/mediapipe/blob/31364660a65fd6f7aa5014c2c87adfce0396911e/mediapipe/web/graph_runner/graph_runner_llm_inference_lib.ts#L118) is not open source?

### Programming Language and version (e.g. C++, Python, Java)

TypeScript and C++

### Describe the current behaviour/state

The MediaPipe GenAI API currently only exposes:

* `generateResponse()` – returns the final text
* `generateResponses()` – returns multiple text responses
* `sizeInTokens()` – returns the token count

There is no way to access the model’s raw logits or log probabilities for each token.

### Describe the the future state

Add an API function, e.g. `logitsForText(query: Prompt): Promise<Logits2D>` in [`llm_inference.ts`](https://github.com/google-ai-edge/mediapipe/blob/master/mediapipe/tasks/web/genai/llm_inference/llm_inference.ts), that computes and returns the model’s raw output scores (logits) for each token in a text prompt.

```
type Logits2D = Float32Array[]; // [T][V]
```

Where:

* **T** = number of tokens in the input
* **V** = vocabulary size
* `logits[t][v]` = unnormalized logit for token *v* at position *t*

### Will this change the current API? How?

Yes, an additional function, e.g. `logitsForText(query: Prompt): Promise<Logits2D>`, needs to be added to [`llm_inference.ts`](https://github.com/google-ai-edge/mediapipe/blob/master/mediapipe/tasks/web/genai/llm_inference/llm_inference.ts).

Furthermore, an additional function needs to be added to [WasmLlmInferenceModule](https://github.com/google-ai-edge/mediapipe/blob/31364660a65fd6f7aa5014c2c87adfce0396911e/mediapipe/web/graph_runner/graph_runner_llm_inference_lib.ts#L118). 

### Please specify the use cases for this feature

Access to logits or log probabilities enables:

* Token-level scoring and likelihood estimation
* Custom decoding and sampling strategies
* Token-level confidence or uncertainty
* Classification via likelihood, e.g. like the [Google Showcase for Gemma Fine-Tuning](https://ai.google.dev/gemma/docs/agile_classifiers)

It a basic functionality of LLMs, so it should be there :)

**Comment by HarishPermal** [comment]
Hi @fosple,

Apologies for the delayed response,

Thanks for raising this issue, i'll discuss this with the team and let you know with updates. 😊

**Comment by fosple** [comment]
Thanks for the reply @HarishPermal 

And two follow up questions, as sometimes one needs a bit more access to the raw ML models (which MediaPipe does not provide because I understand it's more about making it easy to use AI tasks rather than giving full access):

1. Are there any plans to open source [WasmLlmInferenceModule](https://github.com/google-ai-edge/mediapipe/blob/31364660a65fd6f7aa5014c2c87adfce0396911e/mediapipe/web/graph_runner/graph_runner_llm_inference_lib.ts#L118) or any other solution which allows me to access Gemma models more fully in the browser at some point in time? 

2. After a bit of researching my guess is, that WasmLlmInferenceModule is build on [LiteRT-LM](https://github.com/google-ai-edge/LiteRT-LM) and [LiteRT.js](https://github.com/google-ai-edge/LiteRT/tree/main/litert/js): is that correct?

Btw, super cool what you build with MediaPipe - keep on the good work. For now it looks like Transformers.js is the only way forward if I need a bit deeper access to the raw models.

**Comment by tyrmullen** [comment]
To answer your two questions above:

1.  I'm personally hoping to open source more of our MediaPipe Web C++ code and BUILD rules/macros, just for the sake of transparency and collaboration, but I don't think doing so really becomes truly usable until we have open-sourced our GPU backend (a.k.a. "ML Drift" or "MLDrift" in some of our blogs and papers). Otherwise, building any of our Tasks/Solutions for GPU on web will not work, since everything eventually relies on the backend for ML operation support and inference. Allowing for dynamic Wasm linking is an intermediate possibility we've considered, but there are currently some technical challenges with our infrastructure which we'd have to overcome first.

2. While the information online may be a bit confusing on this point, no, they are 3 rather separate projects right now, although the hope is for that to not be the case in the future. At the very least, we'd like for LiteRT.js and LiteRT-LM to be able to use most of the same underlying LiteRT systems ASAP, but trying to merge the LLM efforts is a much bigger undertaking. 

LiteRT-LM does not run on web at all yet, and LiteRT.js cannot load larger model files (like Gemma 3n), so most of our web LLMs need to be run in a completely different way (and hence our "-Web" models are in a very different format from our other models).

I should also mention that even though MediaPipe's "usual" web inference pipeline cannot run AI EdgeTorch-converted .task files (like the ones you would get if you fine-tuned a Gemma 3 1B or Gemma 3 270M model), MediaPipe LLM Inference API currently will detect such files automatically and instead try to run them using pieces from the new systems instead. But this comes with some *significant* performance and behavior/functionality penalties, so the "-web" models are strongly preferable.

TL;DR: They are 3 very different projects, but MediaPipe LLM Inference API is our only option for LLMs on web right now, and although it supports a few different model types, I'd strongly recommend using "-web.*" models whenever possible.

And thank you!!

**Comment by fosple** [comment]
Thanks a lot for the super detailed answer @tyrmullen ! That really makes the current efforts of Google in the AI space way more clear. Already looking forward to the future releases in the different projects :)

**Comment by HarishPermal** [comment]
Hi @tyrmullen,
[This](https://github.com/google-ai-edge/mediapipe/issues/6124#issuecomment-3458246332) is a great explanation 😊, thank you for taking the time to write it up!

@fosple, I hope this helps answer your question. If you're all set, please feel free to close this issue. Let us know if you have any other questions! 😉
